File: api-pytest-framework/agents/prompt_store.py
# ============================================================
# Prompt Store — Stockage et versioning des prompts LLM
# ============================================================
# Chaque prompt est stocké dans prompts/<name>.json avec :
#   - version active (current)
#   - historique complet des versions
#   - métriques collectées (calls, avg_confidence)
#
# Utilisé par les agents pour charger leurs prompts :
#   from prompt_store import PromptStore
#   store = PromptStore()
#   prompt = store.get("triage_classify")  → version active
# ============================================================

import os, json, re
from datetime import datetime

FRAMEWORK    = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
PROMPTS_DIR  = os.path.join(FRAMEWORK, "prompts")

os.makedirs(PROMPTS_DIR, exist_ok=True)


def _now() -> str:
    return datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%SZ")

def _bump(version: str, part: str = "patch") -> str:
    """Incrémente un numéro de version semver (major.minor.patch)."""
    parts = [int(x) for x in version.split(".")]
    while len(parts) < 3:
        parts.append(0)
    if part == "major": parts[0] += 1; parts[1] = 0; parts[2] = 0
    elif part == "minor": parts[1] += 1; parts[2] = 0
    else: parts[2] += 1
    return ".".join(str(p) for p in parts)


class PromptStore:

    def __init__(self, prompts_dir: str = PROMPTS_DIR):
        self.dir = prompts_dir
        os.makedirs(self.dir, exist_ok=True)

    def _path(self, name: str) -> str:
        return os.path.join(self.dir, f"{name}.json")

    def _load(self, name: str) -> dict | None:
        p = self._path(name)
        if not os.path.exists(p):
            return None
        return json.load(open(p, encoding="utf-8"))

    def _save(self, name: str, data: dict):
        with open(self._path(name), "w", encoding="utf-8") as f:
            json.dump(data, f, ensure_ascii=False, indent=2)

    # ── Lecture ────────────────────────────────────────────────────────

    def get(self, name: str, version: str = None) -> str | None:
        """Retourne le contenu du prompt (version active si version=None)."""
        data = self._load(name)
        if not data:
            return None
        if version is None:
            version = data["current_version"]
        for entry in data["history"]:
            if entry["version"] == version:
                return entry["content"]
        return None

    def get_meta(self, name: str) -> dict | None:
        """Retourne les métadonnées complètes du prompt."""
        return self._load(name)

    def list_all(self) -> list:
        """Liste tous les prompts disponibles avec leur version active."""
        result = []
        for f in sorted(os.listdir(self.dir)):
            if f.endswith(".json"):
                data = json.load(open(os.path.join(self.dir, f), encoding="utf-8"))
                result.append({
                    "name":            data["name"],
                    "current_version": data["current_version"],
                    "description":     data.get("description",""),
                    "agent":           data.get("agent",""),
                    "nb_versions":     len(data["history"]),
                    "metrics":         data.get("metrics", {}),
                })
        return result

    def list_versions(self, name: str) -> list:
        """Liste l'historique des versions d'un prompt."""
        data = self._load(name)
        if not data:
            return []
        current = data["current_version"]
        return [
            {**e, "is_current": e["version"] == current}
            for e in data["history"]
        ]

    # ── Création / mise à jour ─────────────────────────────────────────

    def create(self, name: str, content: str, description: str = "",
               agent: str = "", tags: list = None) -> str:
        """Crée un nouveau prompt (version 1.0.0). Erreur si déjà existant."""
        if self._load(name):
            raise ValueError(f"Prompt '{name}' existe déjà — utilise save() pour une nouvelle version.")
        version = "1.0.0"
        data = {
            "name":            name,
            "description":     description,
            "agent":           agent,
            "tags":            tags or [],
            "current_version": version,
            "metrics":         {"calls": 0, "avg_confidence": None, "last_used": None},
            "history": [{
                "version":    version,
                "content":    content,
                "created_at": _now(),
                "note":       "Version initiale",
            }]
        }
        self._save(name, data)
        return version

    def save(self, name: str, content: str, note: str = "",
             bump: str = "patch") -> str:
        """Sauvegarde une nouvelle version (auto-bump semver).
        bump = 'major' | 'minor' | 'patch'
        """
        data = self._load(name)
        if not data:
            return self.create(name, content, note)

        new_version = _bump(data["current_version"], bump)
        data["history"].append({
            "version":    new_version,
            "content":    content,
            "created_at": _now(),
            "note":       note or f"Auto-save depuis bump={bump}",
        })
        data["current_version"] = new_version
        self._save(name, data)
        return new_version

    def promote(self, name: str, version: str):
        """Définit une version existante comme version active."""
        data = self._load(name)
        if not data:
            raise ValueError(f"Prompt '{name}' introuvable.")
        versions = [e["version"] for e in data["history"]]
        if version not in versions:
            raise ValueError(f"Version {version} introuvable. Disponibles : {versions}")
        data["current_version"] = version
        self._save(name, data)

    def rollback(self, name: str) -> str:
        """Revient à la version précédente."""
        data = self._load(name)
        if not data or len(data["history"]) < 2:
            raise ValueError(f"Pas de version précédente pour '{name}'.")
        current = data["current_version"]
        history = data["history"]
        for i, e in enumerate(history):
            if e["version"] == current and i > 0:
                prev = history[i - 1]["version"]
                data["current_version"] = prev
                self._save(name, data)
                return prev
        raise ValueError("Impossible de trouver la version précédente.")

    # ── Métriques ──────────────────────────────────────────────────────

    def record_usage(self, name: str, confidence: float = None):
        """Enregistre un usage du prompt (appelé par les agents)."""
        data = self._load(name)
        if not data:
            return
        m = data.setdefault("metrics", {"calls": 0, "avg_confidence": None, "last_used": None})
        m["calls"] = m.get("calls", 0) + 1
        m["last_used"] = _now()
        if confidence is not None:
            prev = m.get("avg_confidence") or confidence
            n    = m["calls"]
            m["avg_confidence"] = round((prev * (n - 1) + confidence) / n, 3)
        self._save(name, data)

File: perf_k6_qacart/scripts/agents/shared/prompt-store.js
'use strict';
// ============================================================
// Prompt Store — Versioning git-like des prompts LLM
// Stocke chaque prompt dans prompts/<name>.json
// Semver auto-bump : 1.0.0 → 1.0.1 (patch) → 1.1.0 (minor) → 2.0.0 (major)
// ============================================================
const fs   = require('fs');
const path = require('path');

const FRAMEWORK   = path.join(__dirname, '..', '..', '..');
const PROMPTS_DIR = path.join(FRAMEWORK, 'prompts');

fs.mkdirSync(PROMPTS_DIR, { recursive: true });

function nowStr() { return new Date().toISOString().replace(/\.\d+Z$/, 'Z'); }

function bumpVersion(v, part = 'patch') {
  const p = v.split('.').map(Number); while (p.length < 3) p.push(0);
  if (part === 'major')      { p[0]++; p[1] = 0; p[2] = 0; }
  else if (part === 'minor') { p[1]++; p[2] = 0; }
  else                       { p[2]++; }
  return p.join('.');
}

function pPath(name)  { return path.join(PROMPTS_DIR, `${name}.json`); }
function pLoad(name)  { try { return JSON.parse(fs.readFileSync(pPath(name), 'utf8')); } catch { return null; } }
function pSave(name, data) { fs.writeFileSync(pPath(name), JSON.stringify(data, null, 2), 'utf8'); }

// ── Création (v1.0.0) ──────────────────────────────────────────────────────────
function create(name, content, { description = '', agent = '', tags = [] } = {}) {
  if (pLoad(name)) throw new Error(`Prompt '${name}' existe déjà — utilise saveVersion()`);
  const data = {
    name, description, agent, tags,
    current_version: '1.0.0',
    metrics: { calls: 0, avg_confidence: null, last_used: null },
    history: [{ version: '1.0.0', content, created_at: nowStr(), note: 'Version initiale' }],
  };
  pSave(name, data); return '1.0.0';
}

// ── Nouvelle version (auto-bump) ───────────────────────────────────────────────
function saveVersion(name, content, note = '', bump = 'patch') {
  const data = pLoad(name);
  if (!data) return create(name, content, { description: note });
  const nv = bumpVersion(data.current_version, bump);
  data.history.push({ version: nv, content, created_at: nowStr(), note: note || `bump=${bump}` });
  data.current_version = nv;
  pSave(name, data); return nv;
}

// ── Lecture ────────────────────────────────────────────────────────────────────
function get(name, version = null) {
  const data = pLoad(name); if (!data) return null;
  const v = version || data.current_version;
  return data.history.find(e => e.version === v)?.content || null;
}

function getMeta(name) { return pLoad(name); }

function listAll() {
  if (!fs.existsSync(PROMPTS_DIR)) return [];
  return fs.readdirSync(PROMPTS_DIR)
    .filter(f => f.endsWith('.json'))
    .map(f => { const d = JSON.parse(fs.readFileSync(path.join(PROMPTS_DIR, f), 'utf8'));
      return { name: d.name, current_version: d.current_version, description: d.description, nb_versions: d.history.length, metrics: d.metrics }; });
}

function listVersions(name) {
  const data = pLoad(name); if (!data) return [];
  return data.history.map(e => ({ ...e, is_current: e.version === data.current_version }));
}

// ── Gestion versions ───────────────────────────────────────────────────────────
function promote(name, version) {
  const data = pLoad(name); if (!data) throw new Error(`Prompt '${name}' introuvable`);
  if (!data.history.find(e => e.version === version)) throw new Error(`Version ${version} introuvable`);
  data.current_version = version; pSave(name, data);
}

function rollback(name) {
  const data = pLoad(name);
  if (!data || data.history.length < 2) throw new Error(`Pas de version précédente pour '${name}'`);
  const idx = data.history.findIndex(e => e.version === data.current_version);
  if (idx <= 0) throw new Error('Impossible de rollback');
  data.current_version = data.history[idx - 1].version;
  pSave(name, data); return data.current_version;
}

// ── Métriques ──────────────────────────────────────────────────────────────────
function recordUsage(name, confidence = null) {
  const data = pLoad(name); if (!data) return;
  data.metrics.calls = (data.metrics.calls || 0) + 1;
  data.metrics.last_used = nowStr();
  if (confidence !== null) {
    const n = data.metrics.calls, prev = data.metrics.avg_confidence || confidence;
    data.metrics.avg_confidence = Math.round((prev * (n-1) + confidence) / n * 1000) / 1000;
  }
  pSave(name, data);
}

module.exports = { create, saveVersion, get, getMeta, listAll, listVersions, promote, rollback, recordUsage, PROMPTS_DIR };
